install.sh

```diff
@@ -296,176 +293,309 @@ int main() {
-# Write config
+  if [ -z "${CHOSEN_INSTALL_MODE:-}" ]; then
+    echo "Sandstorm makes it easy to run web apps on your own server. You can have:"
+    echo ""
+    echo "1. A full server with automatic setup (press enter to accept this default)"
+    echo "2. A development server, for writing apps."
+    echo ""
+    CHOSEN_INSTALL_MODE=$(prompt "How are you going to use this Sandstorm install?" "1")
+  fi
 
-writeConfig() {
-  while [ $# -gt 0 ]; do
-    eval echo "$1=\$$1"
-    shift
-  done
+  if [ "1" = "$CHOSEN_INSTALL_MODE" ] ; then
+    full_server_install
+  else
+    dev_server_install
+  fi
 }
 
-# TODO(someday): Ask what channel to use. Currently there is only one channel.
-CHANNEL=dev
+dev_server_install() {
+  # Use these settings for a dev-server-oriented install.
+  #
+  # Users will find themselves going through this flow if they
+  # manually choose a dev-server-related flow, but also if they pass
+  # -d on the command line. (The Vagrantfile and the test suite both
+  # use -d. The test suite runs install.sh with -d -u.)
+  #
+  # A "dev server install" must be run as root, unless you pass
+  # -u. That's because app development (aka spk dev) requires running
+  # as root, at the moment.
+
+  if [ "yes" = "$PREFER_ROOT" ] && [ "no" = "$CURRENTLY_UID_ZERO" ] ; then
+    # We are not root, but we would like to be root.
+    echo "If you want app developer mode for a Sandstorm install, you need root"
+    echo "due to limitations in the Linux kernel."
+    echo ""
+
+    echo "To set up Sandstorm, we will need to use sudo."
+    echo "Rest assured that Sandstorm itself won't run as root."
+
+    # If we are running in USE_DEFAULTS mode, then it is not OK to ask
+    # for permission to use sudo.
+    if [ "yes" = "$USE_DEFAULTS" ] ; then
+      ACCEPTED_SUDO_FOR_DEV_SERVER="no"
+    else
+      if prompt-yesno "OK to continue?" "yes" ; then
+        ACCEPTED_SUDO_FOR_DEV_SERVER="yes"
+      else
+        ACCEPTED_SUDO_FOR_DEV_SERVER="no"
+      fi
+    fi
 
-if [ -e sandstorm.conf ]; then
-  echo "Found existing sandstorm.conf. Using it."
-  . sandstorm.conf
-  if [ "${SERVER_USER:+set}" != set ]; then
-    fail "Existing config does not set SERVER_USER. Please fix or delete it."
+    if [ "yes" = "$ACCEPTED_SUDO_FOR_DEV_SERVER" ] ; then
+      rerun_script_as_root CHOSEN_INSTALL_MODE=2
+    else
+      # Print a message that allows people to make an informed decision.
+      echo "If you are OK with a local Sandstorm install for testing"
+      echo "but not app development, re-run install.sh with -u to bypass this message."
+      fail "For developer mode to work, the script needs root, or read above to bypass."
+    fi
   fi
-  if [ "${UPDATE_CHANNEL:-none}" != none ]; then
-    CHANNEL=$UPDATE_CHANNEL
+
+  # If they did not pass -d, then let them opt into that.
+  if [ "yes" != "$USE_DEFAULTS" ] ; then
+    echo "We're going to:"
+    echo ""
+    echo "* Install Sandstorm in ${DEFAULT_DIR_FOR_ROOT}."
```

Comment:
If the script executed with `-u`, then you need DEFAULT_DIR_FOR_NON_ROOT here. (Also for the live server path.)